lib/api/libhoney.js

```diff
@@ -205,10 +201,30 @@ module.exports = class LibhoneyEventAPI {
         [schema.BEELINE_VERSION]: pkg.version,
         [schema.NODE_VERSION]: process.version,
       });
-      if (this.ds) {
-        ev.sampleRate = this.ds.sampleRate;
+
+      let presampled = false;
+      if (this.samplerHook) {
+        const samplerHookResponse = this.samplerHook(ev.data);
+        const keepEvent = samplerHookResponse.shouldSample;
+        if (!keepEvent) {
+          debug("skipping event due to sampler hook sample ev = " + JSON.stringify(ev.data));
+          return;
+        }
+
+        const newSampleRate = samplerHookResponse.newSampleRate;
+        if (newSampleRate !== undefined && typeof newSampleRate === "number") {
+          ev.sampleRate = samplerHookResponse.newSampleRate;
+        }
+        presampled = true;
+      }
+
+      if (presampled) {
+        debug("enqueuing presampled event ev = " + JSON.stringify(ev.data));
```

Comment:
I don't think we need to stringify the span data twice per `finishSpan`.  Either remove the stringify here or from the top of this function.

we're also using new enough nodes that we should use interpolated strings here:
```suggestion
        debug(`enqueuing event ${JSON.stringify(ev)}`);
```
(^ note `ev` instead of `ev.data` so that it also logs the sample rate.)

Reply:
Cool cool. I usually work in Babel so my 'definition' of node is fairly confused.

Reply:
What I've done for now is extract the `stringify` action itself to a variable so at least we aren't re-stringifying every time.

I copied the logging strategy from the `beeline-python` library. Had found having this much debug very helpful when getting up and running. Given these logs will only be called if the debug mode is enabled, could we possibly keep them in?